### ISSUE TITLE
App fails to boot because Maily expects the asset pipeline to be present

### ISSUE THREAD

**Issue by barrywoolgar** [markets/maily#46]
I don't know if this constitutes a bug, but my Rails 6 app is webpacker-only and after installing the gem it is not possible to enter the rails console (or run the generator). Consider this a half bug report, half workaround guide for anyone who needs it.

The bug part is because `lib/maily/engine.rb` attempts to modify `config.assets.precompile`, which does not exist without the asset pipeline and therefore raises an `undefined method 'assets' for #<Rails::Engine::Configuration ...>` error.

If I wrap that section in a `config.respond_to?(:assets)` then everything works, except the assets, as you would expect. I've just copied the SCSS out into a new pack file, and overridden the `maily/application` layout file in my app. Now everything except the images work, and I'm fine with that.

I'm not sure what the proper solution is. The first thing that comes to mind is maybe include a static version of the compiled assets in the gem and update the templates to source them out of `/public` using something like:

```
module Maily
  class Engine < ::Rails::Engine
    ...
    config.app_middleware.use(
      Rack::Static,
      urls: ["/maily-assets"], root: "maily/public"
    )
  end
end
```

Happy to do a PR with the `respond_to?` bit in, but that seems like only half the job.

Thanks

**Comment by markets** [comment]
Hi @barrywoolgar and thanks for such a detailed description!

I usually use this gem in several apps with `webpacker` with no problem, but that's because I still use `sprockets` in combination (js -> webpack, css+images -> sprockets), so we don't faced the problem yet. 

There is no "clear" way to use `webpacker` in Rails engines currently (https://github.com/rails/webpacker/blob/master/docs/engines.md), and at the same time keep compatibility with the `sprockets` version. From what I read in different issues and see in other gems with assets, it's possible, but still tricky and harder to maintain.

So I'd say, we have 2 feasible options to solve this situation:

a) as you pointed out above, provide "static" versions of the assets and inject a middleware to render those ones. 
b) make `maily` depend on `sprocket-rails` (and load `require "sprockets/railtie"`), so `config.assets` will we available.

I think the first option `a` will be nice (no extra deps and we can cut `sass` dep too) but requires some analysis, work and testing. I'd try to go with `b` for now (If I'm not wrong, `sprocket-rails` is still a Rails dep), so probably you only need to: `require "sprockets/railtie"` in `config/application.rb`.

Maybe in the future, if finally Rails settles to one pipeline approach, this will be easier for engines too.

**Comment by barrywoolgar** [comment]
Hi @markets all that makes a lot of sense, thank you for the rapid response. 

You're right, I had the Sprockets railtie commented out. However, this happens by default when the project is initialised with `rails new my_app --skip-sprockets` so I may not be the only one to run into this.

As such, I do think it's worth allowing for the absence of `config.assets` - and maybe displaying a warning when the generator is run? - so that just installing the gem doesn't risk "breaking" a streamlined app just by being in the Gemfile. 

It's definitely an edge case though (and now this issue exists for anyone searching with the same exception), so I'd be happy to close this issue - unless you'd like a PR to address the boot crash issue?

**Comment by markets** [comment]
Yes, I think it's also commented out when you create a Rails API project.

> As such, I do think it's worth allowing for the absence of config.assets - and maybe displaying a warning when the generator is run? - so that just installing the gem doesn't risk "breaking" a streamlined app just by being in the Gemfile.

Yeah! +1 to improve user experience and guidance for a friendly installation! `--skip-sprockets` and `--api` probably aren't not so rare these days.

**Comment by barrywoolgar** [comment]
Hello, apologies for the delay in replying. I'm juggling some things at the moment, but I haven't forgotten!

**Comment by markets** [comment]
Closed by #48
